### PR TITLE
[#323] added model test

### PR DESCRIPTION
**Pull request by paromita92**

### CODE REVIEW

File: spec/support/factory_girl.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# spec/support/factory_girl.rb
+RSpec.configure do |config|
+  config.include FactoryGirl::Syntax::Methods
+end
+
```
Comment (nbanerjee) [inline spec/support/factory_girl.rb]: Remove unwanted blank line

File: config/environments/test.rb [hunk 2]
```diff
@@ -33,4 +33,9 @@
 
   # Raises error for missing translations
   # config.action_view.raise_on_missing_translations = true
+
+
```
Comment (nbanerjee) [inline config/environments/test.rb]: Remove unwanted blank line

File: spec/models/brand_spec.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+require 'rails_helper'
+
+RSpec.describe Brand, type: :model do
+  describe "#provides category wise item count" do
+    let!(:category1) { create(:category) }
+    let!(:category2) { create(:category) }
+    let!(:brand1) { create(:brand) }
+    let!(:brand2) { create(:brand) }
+    let!(:item1) { create(:item, category: category1, brand: brand1) }
+    let!(:item2) { create(:item, category: category2, brand: brand2) }
```
Comment (nbanerjee) [inline spec/models/brand_spec.rb]: Please create whatever you needs.

File: spec/models/brand_spec.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    context "provides the number of item's within a particular brand"
+      it "provides one count of category" do
+        expect(brand1.category_wise_items_count).to eq("#{category1.name}-1")
+      end
+
+    context "do not provide the number of item's within a particular brand"
+      it "do not provide count of category" do
+        expect(brand1.category_wise_items_count).to be_nil
+    end
+  end
```
Comment (nbanerjee) [inline spec/models/brand_spec.rb]: plz put end in context end. And it's not clear what you want to test here.

File: spec/models/brand_spec.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
+require 'rails_helper'
+
+RSpec.describe Brand, type: :model do
+  describe "#provides category wise item count" do
+    let!(:category) { create(:category) }
+    let!(:brand) { create(:brand) }
+    let!(:item) { create(:item, category: category, brand: brand) }
+
+    context "the number of item's for different brand categorie's"
```
Comment (nbanerjee) [inline spec/models/brand_spec.rb]: context should be `When item present / When item not present`

File: spec/models/brand_spec.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+      it "should return one item for a category of brand " do
+        expect(brand.category_wise_items_count).to eq("#{category.name}-1")
+      end
```
Comment (nbanerjee) [inline spec/models/brand_spec.rb]: Still end is missing

File: spec/models/brand_spec.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+    context "no item's for different brand categorie's"
+      it "should not return anything" do
+        expect(brand.category_wise_items_count).to be_nil
+    end
```
Comment (nbanerjee) [inline spec/models/brand_spec.rb]: This test should not pass

File: spec/factories/items.rb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+require 'faker'
+
+FactoryGirl.define do
+  factory :item do |f|
+   f.category { Faker::Name.category }
+   f.brand { Faker::Name.brand }
```
Comment (nbanerjee) [inline spec/factories/items.rb]: What you want to do here?

File: spec/models/brand_spec.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    let!(:item) { create(:item, category: category, brand: brand) }
+
+    context "when item present" do
+      it "should return one item for a category of brand " do
+        expect(brand.category_wise_items_count).to eq("#{category.name}-1")
+      end
+    end
+
+    context "when item not present" do
+      it "should not return anything" do
```
Comment (nbanerjee) [inline spec/models/brand_spec.rb]: make it should return blank string

File: spec/support/factory_girl.rb [hunk 10]
```diff
@@ -0,0 +1,4 @@
+# spec/support/factory_girl.rb
+RSpec.configure do |config|
+  config.include FactoryGirl::Syntax::Methods
```
Comment (nbanerjee) [inline spec/support/factory_girl.rb]: why it's needed?

File: spec/models/brand_spec.rb [hunk 11]
```diff
@@ -0,0 +1,22 @@
+require 'rails_helper'
+
+RSpec.describe Brand, type: :model do
```
Comment (nbanerjee) [inline spec/models/brand_spec.rb]: ```describe Brand do```

File: spec/spec_helper.rb [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+  #     --seed 1234
+  config.order = :random
+
+  # Seed global randomization in this process using the `--seed` CLI option.
+  # Setting this allows you to use `--seed` to deterministically reproduce
+  # test failures related to randomization by passing the same `--seed` value
+  # as the one that triggered the failure.
+  Kernel.srand config.seed
+=end
+end
```
Comment (nbanerjee) [inline spec/spec_helper.rb]: Remove this file.

File: spec/models/brand_spec.rb [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+describe Brand do
+  describe "#provides category wise item count" do
+    let!(:category)     { create(:category) }
+    let!(:brand)        { create(:brand) }
+    let!(:anotherbrand) { create(:brand) }
+    let!(:item)         { create(:item, category: category, brand: brand) }
+
+    context "when item present" do
+      it "should return one item for a category of brand " do
```
Comment (nbanerjee) [inline spec/models/brand_spec.rb]: It's not clear

File: spec/models/checkout_spec.rb [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+      it "should return false" do
+        expect(another_checkout.checkin?).to be false
+      end
+    end
+  end
+
+  describe "#checkout_limitation" do
+    context "when checkout date before purchase date" do
+      let!(:item)     { create(:item, purchase_on: (Date.today)) }
+      let!(:checkout) { build(:checkout, item: item, checkout: (Date.today-1)) }
```
Comment (nbanerjee) [inline spec/models/checkout_spec.rb]: Add new line

File: spec/factories/resolutions.rb [hunk 15]
```diff
@@ -0,0 +1,6 @@
+require 'faker'
+FactoryGirl.define do
+  factory :resolution do |f|
+    f.name{Faker::Name.name}
```
Comment (sandeepc19) [inline spec/factories/resolutions.rb]: give space before and after `{` and `}`

File: spec/factories/vendors.rb [hunk 16]
```diff
@@ -0,0 +1,7 @@
+require 'faker'
+FactoryGirl.define do
+  factory :vendor do |f|
+    f.name{Faker::Name.name}
```
Comment (sandeepc19) [inline spec/factories/vendors.rb]: give space before and after `{` and `}`

File: spec/models/brand_spec.rb [hunk 17]
```diff
@@ -0,0 +1,47 @@
+require 'rails_helper'
+
+describe Brand do
+  context "scopes" do
+    let!(:brand)         { create(:brand, deleted_at: nil, name: "Microsoft") }
+    let!(:another_brand) { create(:brand, deleted_at: Time.now, name: "Frontech") }
+
+    describe ".active" do
+      context "when brand is active" do
+        it "deleted_at should be nil" do
```
Comment (sandeepc19) [inline spec/models/brand_spec.rb]: write proper it condition

File: spec/models/brand_spec.rb [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    end
+
+    describe ".order_by_name" do
+      it "should order brand names alphabetically" do
+        expect(Brand.order_by_name).to eq([another_brand, brand])
+      end
+    end
+  end
+
+  describe "#provides category wise item count" do
```
Comment (sandeepc19) [inline spec/models/brand_spec.rb]: write exact method name

File: spec/models/category_spec.rb [hunk 19]
```diff
@@ -0,0 +1,28 @@
+require "rails_helper"
+
+describe Category do
+  context "scopes" do
+    let!(:category)         { create(:category, deleted_at: nil, name: "Mouse") }
+    let!(:another_category) { create(:category, deleted_at: Time.now, name: "Keyboard") }
+
+    describe ".active" do
+      context "when category is active" do
+        it "deleted_at should be nil" do
```
Comment (sandeepc19) [inline spec/models/category_spec.rb]: change it condition

File: spec/factories/vendor.rb [hunk 20]
```diff
@@ -0,0 +1,7 @@
+require 'faker'
+FactoryGirl.define do
+  factory :vendor do |f|
+    f.name { Faker::Name.name }
```
Comment (sandeepc19) [inline spec/factories/vendor.rb]: fix indentation

File: spec/models/checkout_spec.rb [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+
+    context "when check_in not present" do
+      it "should return false" do
+        expect(another_checkout.checkin?).to be false
+      end
+    end
+  end
+
+  describe "#checkout_limitation" do
+
```
Comment (sandeepc19) [inline spec/models/checkout_spec.rb]: remove this line break

File: spec/models/checkout_spec.rb [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+    context "when checkout date is before purchase date" do
```
Comment (sandeepc19) [inline spec/models/checkout_spec.rb]: give a line break

File: spec/models/checkout_spec.rb [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+      let!(:item)     { create(:item, purchase_on: (Date.today)) }
+      let!(:checkout) { build(:checkout, item: item, checkout: (Date.today-1)) }
```
Comment (sandeepc19) [inline spec/models/checkout_spec.rb]: give a line break

File: spec/models/checkout_spec.rb [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+      it "should be invalid" do
+        expect(checkout).to be_invalid
+      end
+    end
+
+    context "when checkout date is after purchase date" do
```
Comment (sandeepc19) [inline spec/models/checkout_spec.rb]: give line break

File: spec/models/checkout_spec.rb [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+      let!(:another_item)     { create(:item, purchase_on: (Date.today-1)) }
+      let!(:another_checkout) { create(:checkout, item: another_item, checkout: (Date.today)) }
```
Comment (sandeepc19) [inline spec/models/checkout_spec.rb]: give line break

File: spec/models/issue_spec.rb [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+
+    describe ".unclosed" do
+      context "when closed_at is nil" do
+        it "should return unclosed issue" do
+          expect(Issue.unclosed).to include(another_issue)
+        end
+      end
+
+      context "when closed_at is not nil" do
+        it "should not return unclosed issues" do
```
Comment (sandeepc19) [inline spec/models/issue_spec.rb]: make it `it "should not return closed issues"`

File: spec/models/user_spec.rb [hunk 27]
```diff
@@ -1,6 +1,26 @@
...
+
+    describe ".order_by_name" do
+      it "should order User's names alphabetically according to first name" do
+        expect(User.order_by_name).to eq([another_user, user])
+      end
+    end
+  end
+
+  describe "#full_name" do
+    let!(:another_user) { create(:user, first_name: "Aman", last_name: "Arora") }
```
Comment (sandeepc19) [inline spec/models/user_spec.rb]: give line break

File: spec/models/item_spec.rb [hunk 28]
```diff
@@ -0,0 +1,433 @@
...
+        it "dont return those items which are attached" do
+          expect(Item.unattached).not_to include(:item, another_item)
+        end
+      end
+    end
+
+    describe ".unavailable" do
+      let!(:item)         { create(:item, category: category, brand: brand, created_at: Date.today) }
+      let!(:another_item) { create(:item, category: category, brand: brand, created_at: Date.yesterday) }
+      let!(:chkout_1)       { create(:checkout, item: another_item, checkout: Date.today, check_in: Date.tomorrow)}
```
Comment (sandeepc19) [inline spec/models/item_spec.rb]: fix indentation

File: spec/rails_helper.rb [hunk 29]
```diff
@@ -3,8 +3,10 @@
 require File.expand_path('../../config/environment', __FILE__)
 # Prevent database truncation if the environment is production
 abort("The Rails environment is running in production mode!") if Rails.env.production?
+
```
Comment (sandeepc19) [inline spec/rails_helper.rb]: remove extra line

File: spec/models/issue_spec.rb [hunk 30]
```diff
@@ -0,0 +1,48 @@
+require 'rails_helper'
+
+describe Issue do
+
```
Comment (sandeepc19) [inline spec/models/issue_spec.rb]: remove line break

File: spec/models/issue_spec.rb [hunk 31]
```diff
@@ -0,0 +1,48 @@
...
+        it "should not return closed issues" do
+          expect(Issue.unclosed).not_to include(issue)
+        end
+      end
+    end
+  end
+
+  describe "#item_closed_at_limitation" do
+    let!(:item)          { create(:item, purchase_on: Date.today) }
+    let!(:issue)         { build(:issue, item: item, closed_at: Date.today-1) }
```
Comment (sandeepc19) [inline spec/models/issue_spec.rb]: use `create` instead of `build`

File: spec/models/issue_spec.rb [hunk 32]
```diff
@@ -0,0 +1,48 @@
...
+    let!(:another_item)  { create(:item, purchase_on: Date.today-1) }
+    let!(:another_issue) { build(:issue, item: another_item, closed_at: Date.today) }
```
Comment (sandeepc19) [inline spec/models/issue_spec.rb]: use `create` instead of `build`

File: spec/models/user_spec.rb [hunk 33]
```diff
@@ -1,6 +1,28 @@
 require 'rails_helper'
 
 describe User  do
+
```
Comment (sandeepc19) [inline spec/models/user_spec.rb]: remove line break

File: spec/models/vendor_spec.rb [hunk 34]
```diff
@@ -0,0 +1,15 @@
+require "rails_helper"
+
+describe Vendor do
+
```
Comment (sandeepc19) [inline spec/models/vendor_spec.rb]: remove line break

File: spec/models/item_spec.rb [hunk 35]
```diff
@@ -0,0 +1,430 @@
+require 'rails_helper'
+
+describe Item do
+
```
Comment (sandeepc19) [inline spec/models/item_spec.rb]: remove line break

File: spec/models/employee_spec.rb [hunk 36]
```diff
@@ -17,4 +28,17 @@
       end
     end
   end
+
+  describe "#deallocate_items" do
+    let!(:employee) { create(:employee, active: true) }
+    let!(:item)     { create(:item, employee: employee)}
```
Comment (sandeepc19) [inline spec/models/employee_spec.rb]: give space before `}`

File: spec/models/resolution_spec.rb [hunk 37]
```diff
@@ -0,0 +1,15 @@
+require "rails_helper"
+
+describe Resolution do
+
```
Comment (sandeepc19) [inline spec/models/resolution_spec.rb]: remove line break

File: spec/models/user_spec.rb [hunk 38]
```diff
@@ -1,6 +1,27 @@
...
+      it "should order User's names alphabetically according to first name" do
+        expect(User.order_by_name).to eq([another_user, user])
+      end
+    end
+  end
+
+  describe "#full_name" do
+    let!(:user) { create(:user, first_name: "Aman", last_name: "Arora") }
+
+    context "when compiling full name" do
```
Comment (sandeepc19) [inline spec/models/user_spec.rb]: change `combining` to `showing`

File: spec/models/item_spec.rb [hunk 39]
```diff
@@ -0,0 +1,449 @@
...
+        it "should not return the inactive item" do
+          expect(Item.active).not_to include(fourth_item)
+        end
+      end
+    end
+
+    describe ".available" do
+      let!(:item)                  { create(:item, category: category, brand: brand, created_at: Date.today) }
+      let!(:another_item)          { create(:item, category: category, brand: brand, created_at: Date.yesterday) }
+      let!(:another_item_checkout) { create(:checkout, item: another_item, checkout: Date.today, check_in: Date.tomorrow)}
```
Comment (sandeepc19) [inline spec/models/item_spec.rb]: give space before `}`

File: spec/models/item_spec.rb [hunk 40]
```diff
@@ -0,0 +1,449 @@
...
+        it "should not return those items which are attached" do
+          expect(Item.unattached).not_to include(:item, another_item)
+        end
+      end
+    end
+
+    describe ".unavailable" do
+      let!(:item)                  { create(:item, category: category, brand: brand, created_at: Date.today) }
+      let!(:another_item)          { create(:item, category: category, brand: brand, created_at: Date.yesterday) }
+      let!(:another_item_checkout) { create(:checkout, item: another_item, checkout: Date.today, check_in: Date.tomorrow)}
```
Comment (sandeepc19) [inline spec/models/item_spec.rb]: give space before `}`

File: spec/models/item_spec.rb [hunk 41]
```diff
@@ -0,0 +1,449 @@
...
+      let!(:item_checkout)         { create(:checkout, item: item, checkout: Date.tomorrow, check_in: nil)}
```
Comment (sandeepc19) [inline spec/models/item_spec.rb]: give space before `}`

File: spec/models/item_spec.rb [hunk 42]
```diff
@@ -0,0 +1,472 @@
...
+  describe "#change_parent" do
+    let!(:item)         { create(:item) }
+    let!(:second_item)  { create(:item) }
+    let!(:another_item) { create(:item, parent: second_item) }
+
+    context "when parent can be changed" do
+      before do
+        another_item.change_parent(item)
+      end
+      it "should change the parent" do
```
Comment (sandeepc19) [inline spec/models/item_spec.rb]: give line break

File: spec/models/item_spec.rb [hunk 43]
```diff
@@ -0,0 +1,472 @@
...
+
+  describe "#change_parent" do
+    let!(:item)         { create(:item) }
+    let!(:second_item)  { create(:item) }
+    let!(:another_item) { create(:item, parent: second_item) }
+
+    context "when parent can be changed" do
+      before do
+        another_item.change_parent(item)
+      end
```
Comment (sandeepc19) [inline spec/models/item_spec.rb]: give line break

File: spec/models/item_spec.rb [hunk 44]
```diff
@@ -0,0 +1,472 @@
...
+      it "should change the parent" do
+        expect(another_item.parent_id).to eq(item.id)
+      end
+    end
+    context "when parent cannot be changed" do
+      before do
+        second_item.change_parent(another_item)
+      end
```
Comment (sandeepc19) [inline spec/models/item_spec.rb]: give line break

File: spec/models/item_spec.rb [hunk 45]
```diff
@@ -0,0 +1,472 @@
...
+  describe "#edit_item_details" do
+    let!(:category)         { create(:category) }
+    let!(:brand)            { create(:brand) }
+    let!(:another_category) { create(:category) }
+    let!(:another_brand)    { create(:brand) }
+    let!(:item)             { create(:item, category: category, brand: brand, created_at: 6.day.ago.to_datetime) }
+
+    context "when category is changed" do
+      it "should be invalid" do
+        item.update(category: another_category)
```
Comment (sandeepc19) [inline spec/models/item_spec.rb]: write this in `before`

File: spec/models/item_spec.rb [hunk 46]
```diff
@@ -0,0 +1,472 @@
...
+        expect(item).to be_invalid
+      end
+    end
+
+    context "when brand is changed" do
+      it "should be invalid" do
+        item.update(brand: another_brand)
```
Comment (sandeepc19) [inline spec/models/item_spec.rb]: write this in `before`

File: spec/models/item_spec.rb [hunk 47]
```diff
@@ -0,0 +1,484 @@
...
+    context "when employee ID is changed" do
+      it "should update_item_history" do
+        item.update(employee: another_employee)
+        expect(item.item_histories.size).to eq(2)
+      end
+    end
+
+    context "when working status is changed" do
+      it "should update item_history" do
+        item.update(working: false)
```
Comment (sandeepc19) [inline spec/models/item_spec.rb]: write this in `before`

File: spec/models/item_spec.rb [hunk 48]
```diff
@@ -0,0 +1,484 @@
...
+        expect(item.item_histories.size).to eq(2)
+      end
+    end
+
+    context "when parent is changed" do
+      it "should update item_history" do
+        another_item.update(parent: parent_item)
```
Comment (sandeepc19) [inline spec/models/item_spec.rb]: write this in `before`

**Pull request merged**